src/lib/geoip/client.ts

```diff
@@ -126,40 +133,53 @@ export default class GeoipClient {
 			};
 		}
 
-		const maxmind = sources.find(s => s.provider === 'maxmind' && s.city === best.city);
-
-		if (maxmind?.asn && maxmind?.network) {
-			return {
-				asn: maxmind.asn,
-				network: maxmind.network,
-				normalizedNetwork: maxmind.normalizedNetwork,
-			};
+		for (const source of rankedSources) {
+			if (source.normalizedCity === best.normalizedCity && source?.asn && source?.network) {
+				return {
+					asn: source.asn,
+					network: source.network,
+					normalizedNetwork: source.normalizedNetwork,
+				};
+			}
 		}
 
 		return undefined;
 	}
 
-	private bestMatch (field: keyof LocationInfo, sources: LocationInfoWithProvider[]): LocationInfo {
+	private bestMatch (field: keyof LocationInfo, sources: LocationInfoWithProvider[]): [LocationInfo, LocationInfoWithProvider[]] {
+		const DESC_PRIORITY_OF_PROVIDERS: Provider[] = [ 'ip2location', 'ipmap', 'maxmind', 'ipinfo', 'fastly' ];
 		const filtered = sources.filter(s => s[field]);
-		// Group by the same field value
-		const grouped = Object.values(_.groupBy(filtered, field));
-		// Move items with the same values to the beginning
-		const ranked = grouped.sort((a, b) => b.length - a.length).flat();
+		// Initially sort sources so they are sorted inside groups
+		const sorted = filtered.sort((sourceA, sourceB) => {
+			const indexSourceA = DESC_PRIORITY_OF_PROVIDERS.indexOf(sourceA.provider);
+			const indexSourceB = DESC_PRIORITY_OF_PROVIDERS.indexOf(sourceB.provider);
+			return indexSourceA - indexSourceB;
+		});
+		// Group sources by the same field value
+		const grouped = Object.values(_.groupBy(sorted, field));
+		const ranked = grouped.sort((sourcesA, sourcesB) => {
+			// Move bigger groups to the beginning
+			const groupsSizeDiff = sourcesB.length - sourcesA.length;
 
-		let best = ranked[0];
+			if (groupsSizeDiff === 0) {
+				// If group sizes are equal move the group with the prioritized item to the beginning
+				const smallestIndexSourceA = DESC_PRIORITY_OF_PROVIDERS.indexOf((sourcesA[0] as LocationInfoWithProvider).provider);
+				const smallestIndexSourceB = DESC_PRIORITY_OF_PROVIDERS.indexOf((sourcesA[0] as LocationInfoWithProvider).provider);
+				return smallestIndexSourceA - smallestIndexSourceB;
+			}
 
-		// If all values are different
-		if (grouped.length === filtered.length) {
-			const sourcesObject = Object.fromEntries(filtered.map(s => [ s.provider, s ]));
-			best = sourcesObject['ipinfo'] ?? sourcesObject['maxmind'];
-		}
+			return groupsSizeDiff;
+		}).flat();
```

Comment:
Same here. It seems you just wrote precisely what `const ranked = grouped.sort((a, b) => b.length - a.length).flat();` did already, but in a more complicated way.